File: src/main/java/de/krall/reflare/FlareBorder.java
package de.krall.reflare;

import de.krall.flare.dom.Element;
import de.krall.reflare.t.TInsets;
import de.krall.reflare.ui.FlareUI;
import java.awt.Component;
import java.awt.Graphics;
import java.awt.Insets;
import javax.swing.AbstractButton;
import javax.swing.JMenuBar;
import javax.swing.JToolBar;
import javax.swing.border.AbstractBorder;
import javax.swing.text.JTextComponent;

public class FlareBorder extends AbstractBorder {

    private transient FlareUI ui;

    public FlareBorder(final FlareUI ui) {
        this.ui = ui;
    }

    @Override
    public void paintBorder(final Component c, final Graphics g, final int x, final int y, final int width, final int height) {
        ui.paintBorder(c, g, x, y, width, height);
    }

    @Override
    public Insets getBorderInsets(final Component c, Insets insets) {
        if (insets == null) {
            insets = new Insets(0, 0, 0, 0);
        } else {
            insets.top = 0;
            insets.right = 0;
            insets.bottom = 0;
            insets.left = 0;
        }

        final Element element = ui.getElement();

        if (c instanceof JTextComponent) {
            combine(insets, ((JTextComponent) c).getMargin());
        } else if (c instanceof AbstractButton) {
            combine(insets, ((AbstractButton) c).getMargin());
        } else if (c instanceof JToolBar) {
            combine(insets, ((JToolBar) c).getMargin());
        } else if (c instanceof JMenuBar) {
            combine(insets, ((JMenuBar) c).getMargin());
        }

        return insets;
    }

    private void combine(final Insets insets, final Insets addition) {
        if (addition != null) {
            insets.top += addition.top;
            insets.right += addition.right;
            insets.bottom += addition.bottom;
            insets.left += addition.left;
        }
    }

    private void combine(final Insets insets, final TInsets addition) {
        if (addition != null) {
            insets.top += addition.top;
            insets.right += addition.right;
            insets.bottom += addition.bottom;
            insets.left += addition.left;
        }
    }

    @Override
    public boolean isBorderOpaque() {
        return false;
    }
}

File: src/main/java/org/fernice/reflare/ui/FlareBorder.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package org.fernice.reflare.ui

import org.fernice.flare.style.ComputedValues
import org.fernice.reflare.resource.RequiresResourceContext
import org.fernice.reflare.resource.ResourceContext
import org.fernice.reflare.resource.TInsets
import org.fernice.reflare.resource.plus
import org.fernice.reflare.resource.toAWTInsets
import org.fernice.reflare.resource.toTInsets
import org.fernice.reflare.resource.withResourceContext
import java.awt.Component
import java.awt.Graphics
import java.awt.Rectangle
import javax.swing.border.AbstractBorder
import javax.swing.text.JTextComponent
import java.awt.Insets as AWTInsets

private val None = AWTInsets(0, 0, 0, 0)

open class FlareBorder protected constructor(private val ui: FlareUI) : AbstractBorder() {

    override fun paintBorder(c: Component?, g: Graphics, x: Int, y: Int, width: Int, height: Int) {
        ui.paintBorder(c ?: ui.element.component, g)
    }

    override fun getBorderInsets(c: Component?, insets: AWTInsets?): AWTInsets {
        val componentInsets = if (c != null) getComponentInsets(c) else None
        return withResourceContext {
            computeInsets { bounds ->
                margin.toTInsets(bounds) + border.toTInsets() + padding.toTInsets(bounds) + componentInsets
            }.toAWTInsets(insets)
        }
    }

    protected open fun getComponentInsets(c: Component): AWTInsets = None

    @RequiresResourceContext
    internal fun getMarginAndBorderTInsets(): TInsets {
        return computeInsets { bounds -> margin.toTInsets(bounds) + border.toTInsets() }
    }

    fun getMarginAndBorderInsets(): AWTInsets = withResourceContext { getMarginAndBorderTInsets().toAWTInsets() }
    fun getMarginInsets(): AWTInsets = withResourceContext { computeInsets { bounds -> margin.toTInsets(bounds) }.toAWTInsets() }
    fun getPaddingInsets(): AWTInsets = withResourceContext { computeInsets { bounds -> padding.toTInsets(bounds) }.toAWTInsets() }

    @RequiresResourceContext
    private inline fun computeInsets(block: ComputedValues.(Rectangle) -> TInsets): TInsets {
        val style = ui.element.getStyle()
        return if (style != null) {
            val bounds = ui.element.component.getBounds(ResourceContext.Rectangle())

            style.block(bounds)
        } else {
            ResourceContext.TInsets()
        }
    }

    override fun isBorderOpaque(): Boolean {
        return false
    }

    companion object {

        @JvmStatic
        fun create(ui: FlareUI): FlareBorder {
            return when (ui.element.component) {
                is JTextComponent -> TextComponentFlareBorder(ui)
                else -> FlareBorder(ui)
            }
        }
    }
}

class TextComponentFlareBorder(ui: FlareUI) : FlareBorder(ui) {

    override fun getComponentInsets(c: Component): AWTInsets {
        return (c as JTextComponent).margin ?: None
    }
}